pkg/retriever/bitswapretriever.go

```diff
@@ -107,15 +109,14 @@ type bitswapRetrieval struct {
 }
 
 // RetrieveFromCandidates retrieves via go-bitswap backed with the given candidates, under the auspices of a fetcher.Fetcher
-func (br *bitswapRetrieval) RetrieveFromCandidates(candidates []types.RetrievalCandidate) (*types.RetrievalStats, error) {
+func (br *bitswapRetrieval) RetrieveFromAsyncCandidates(ayncCandidates types.InboundAsyncCandidates) (*types.RetrievalStats, error) {
 	phaseStartTime := br.clock.Now()
 	// this is a hack cause we aren't able to track bitswap fetches per peer for now, so instead we just create a single peer for all events
 	bitswapCandidate := types.NewRetrievalCandidate(peer.ID(""), br.request.Cid, metadata.Bitswap{})
 	br.events(events.Started(br.request.RetrievalID, phaseStartTime, types.RetrievalPhase, bitswapCandidate))
 
 	// setup the linksystem to record bytes & blocks written -- since this isn't automatic w/o go-data-transfer
 	ctx, cancel := context.WithCancel(br.ctx)
-	defer cancel()
```

Comment:
did you remove this just to get the `cancel()` up immediately after the traversal returns; are we worried that the `br.events()` calls might be non-trivial?

Reply:
moved this to get immediate cancel, so the context is cancelled asap in the go routines above. We want to make sure when RemoveProviders is called below, we don't get any more added afterward